### PR TITLE
[flang][fir] Add external name mangling pass

### PR DESCRIPTION
**Pull request by clementval**
The patch add a new pass called `ExternalNameConversion`. It converts the FIR internal name with the gfortran external name convention. 

The pass is not applied by default in `tco`.

### CODE REVIEW

File: flang/lib/Optimizer/Transforms/ExternalABIConversion.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
+//===- ExternalABIConversion.cpp -- convert name with external ABI ------*-===//
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/ExternalABIConversion.cpp:1]: '-*-' leftover in the emac header ?

**Review by jeanPerier — APPROVED** [review]
Looks great to me. I wondering if this should be run by default by tco or not. Since tco is not a real Fortran driver, I guess it's OK to not run this pass by default.

Regarding the pass name, I am OK with it, the pedantic comment would be that `--external-abi` might be a bit too generic since this is only dealing with the naming part of the ABI and that the target-rewrite pass (declared [here](https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/include/flang/Optimizer/CodeGen/CGPasses.td#L46)) pass is also somehow Fortran ABI related. So `--external-name-abi` might be more descriptive, but wait for @schweitzpgi takes on this before changing it.

File: flang/lib/Optimizer/Transforms/ExternalABIConversion.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Optimizer/Transforms/ExternalABIConversion.cpp:1]: Just a copy-paste from another file. We might want to make a pass on all files in this directory (an maybe others) because the headers are not consistent.

File: flang/lib/Optimizer/Support/InternalNames.cpp [hunk 2]
```diff
@@ -298,3 +298,19 @@ fir::NameUniquer::deconstruct(llvm::StringRef uniq) {
...
+  return (deconstructResult.first == NameKind::PROCEDURE 
+    || deconstructResult.first == NameKind::COMMON)
+    && deconstructResult.second.modules.empty()
+    && !deconstructResult.second.host;
+}
+
+bool fir::NameUniquer::needExternalABIMangling(llvm::StringRef uniquedName) {
+  auto result = fir::NameUniquer::deconstruct(uniquedName);
+  if (result.first == fir::NameUniquer::NameKind::NOT_UNIQUED || 
+      !fir::NameUniquer::isExternalFacingUniquedName(result))
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Support/InternalNames.cpp:313]: I think you can make this
```c++
  return result.first != fir::NameUniquer::NameKind::NOT_UNIQUED && fir::NameUniquer::isExternalFacingUniquedName(result);
```

File: flang/lib/Optimizer/Transforms/ExternalABIConversion.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+#include "flang/Optimizer/Transforms/Passes.h"
+#include "flang/Optimizer/Support/InternalNames.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/OpenACC/OpenACC.h"
+#include "mlir/Dialect/OpenMP/OpenMPDialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+
+using namespace mlir;
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Transforms/ExternalABIConversion.cpp:20]: We try to avoid importing open namespaces.

**Review by schweitzpgi — COMMENTED** [review]
Looks good. As far as what to call this, maybe something with "interoperability" and "name/naming" vs. "external ABI"? to focus on the naming aspect rather than anything about the binary interface.

**Comment by clementval** [comment]
> Looks good. As far as what to call this, maybe something with "interoperability" and "name/naming" vs. "external ABI"? to focus on the naming aspect rather than anything about the binary interface.

Thanks for the review. I renamed most of the ABI into Name and the pass option is now `--external-name-interop`

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 4]
```diff
@@ -179,4 +179,13 @@ def CharacterConversion : Pass<"character-conversion"> {
   ];
 }
 
+def ExternalNameConversion : Pass<"external-name-interop", "mlir::ModuleOp"> {
+  let summary = "Convert name for external interoperability";
+  let description = [{
+    Demangle FIR internal name and mangle them for external interoperability.
+  }];
+  let constructor = "::fir::createExternalNameConversionPass()";
+  let dependentDialects = [ "fir::FIROpsDialect", "mlir::LLVM::LLVMDialect" ];
```
Comment (schweitzpgi) [inline flang/include/flang/Optimizer/Transforms/Passes.td:188]: Don't forget to add OpenMP, OpenACC, if that was your plan here.

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 5]
```diff
@@ -179,4 +179,13 @@ def CharacterConversion : Pass<"character-conversion"> {
...
+}
```
Comment (schweitzpgi) [inline flang/include/flang/Optimizer/Transforms/Passes.td:189]: I think we'll want a pass option so the user can select a mangling scheme. (Fortran compilers each have their own.)

That can be a separate PR, though. Just handling the basic case to be F77 mode compatible is a great start.

**Review by schweitzpgi — APPROVED** [review]
+1

File: flang/include/flang/Optimizer/Transforms/Passes.td (hunk 5 above)
Comment (clementval) [reply flang/include/flang/Optimizer/Transforms/Passes.td:189]: Yeah I initially added the option and since the F77 scheme is the same for all compiler I tested I just removed it for now.

File: flang/include/flang/Optimizer/Transforms/Passes.td (hunk 4 above)
Comment (clementval) [reply flang/include/flang/Optimizer/Transforms/Passes.td:188]: Thanks I'll add them.

**Comment by clementval** [comment]
FIx for #831 and #1003 (exit is a special case)

**Review by jeanPerier — APPROVED** [review]
